src/compiler/checker.ts

```diff
@@ -21542,9 +21649,55 @@ namespace ts {
+                            chain = chainDiagnosticMessages(chain, Diagnostics.No_overload_matches_this_call);
+                        }
+                        const diags = getSignatureApplicabilityError(node, args, last, assignableRelation, CheckMode.Normal, /*reportErrors*/ true, () => chain);
+                        if (diags) {
+                            for (const d of diags) {
+                                if (last.declaration && candidatesForArgumentError.length > 3) {
+                                    addRelatedInfo(d, createDiagnosticForNode(last.declaration, Diagnostics.The_last_overload_is_declared_here));
+                                }
+                                diagnostics.add(d);
+                            }
+                        }
+                        else {
+                            Debug.fail("No error for last overload signature");
+                        }
+                    }
+                    else {
+                        const allDiagnostics: DiagnosticRelatedInformation[][] = [];
+                        let max = 0;
+                        let min = Number.MAX_VALUE;
+                        let minIndex = 0;
+                        let i = 0;
+                        for (const c of candidatesForArgumentError) {
+                            const chain = () => chainDiagnosticMessages(/*details*/ undefined, Diagnostics.Overload_0_of_1_2_gave_the_following_error, i + 1, candidates.length, signatureToString(c));
+                            const diags = getSignatureApplicabilityError(node, args, c, assignableRelation, CheckMode.Normal, /*reportErrors*/ true, chain);
+                            if (diags) {
+                                if (diags.length <= min) {
+                                    min = diags.length;
+                                    minIndex = i;
+                                }
+                                max = Math.max(max, diags.length);
+                                allDiagnostics.push(diags);
+                            }
+                            else {
+                                Debug.fail("No error for 3 or fewer overload signatures");
+                            }
+                            i++;
+                        }
 
-                if (candidateForArgumentError) {
-                    checkApplicableSignature(node, args, candidateForArgumentError, assignableRelation, CheckMode.Normal, /*reportErrors*/ true);
+                        const related = map(
+                            max > 1 ? allDiagnostics[minIndex] : flatten(allDiagnostics),
+                            d => typeof d.messageText === "string" ? (d as DiagnosticMessageChain) : d.messageText);
+                        diagnostics.add(createDiagnosticForNodeFromMessageChain(node, chainDiagnosticMessages(related, Diagnostics.No_overload_matches_this_call)));

```

Comment:
```suggestion
                        const diags = max > 1 ? allDiagnostics[minIndex] : flatten(allDiagnostics);
                        const related = map(
                            diags,
                            d => typeof d.messageText === "string" ? (d as DiagnosticMessageChain) : d.messageText);
                        diagnostics.add(createDiagnosticForNodeFromMessageChain(node, chainDiagnosticMessages(related, Diagnostics.No_overload_matches_this_call), reduceLeft(diags, (acc, d) => acc.concat(d.relatedInformation || []), [] as DiagnosticRelatedInformation[])));
```
?

Reply:
DiagnosticRelatedInformation doesn't have a property named relatedInformation, just Diagnostic. I'll see whether a cast is safe here.

Reply:
The cast in that suggestion is on an empty array, just so it's not inferred as a `never[]` - or were you referencing something else?

Reply:
`acc.concat(d.relatedInformation || [])` : `d: DiagnosticRelatedInformation` has no property `relatedInformation`. Only its subtype `Diagnostic` does.